src/analysis/retail/priest/shadow/modules/talents/PhantasmalPathogen.tsx

```diff
@@ -0,0 +1,83 @@
+import SPELLS from 'common/SPELLS';
+import TALENTS from 'common/TALENTS/priest';
+import Analyzer, { Options, SELECTED_PLAYER } from 'parser/core/Analyzer';
+import { DamageEvent, EventType } from 'parser/core/Events';
+import Events from 'parser/core/Events';
+import { calculateEffectiveDamage } from 'parser/core/EventCalculateLib';
+import Enemies from 'parser/shared/modules/Enemies';
+import EventHistory from 'parser/shared/modules/EventHistory';
+import BoringSpellValueText from 'parser/ui/BoringSpellValueText';
+import ItemDamageDone from 'parser/ui/ItemDamageDone';
+
+import { PHANTASMAL_PATHOGEN_DAMAGE_PER_RANK } from '../../constants';
+
+class PhantasmalPathogen extends Analyzer {
+  static dependencies = {
+    enemies: Enemies,
+    eventHistory: EventHistory,
+  };
+  protected enemies!: Enemies;
+  protected eventHistory!: EventHistory;
+
+  castSA = 0;
+
+  phantasmalPathogenMultiplier =
+    this.selectedCombatant.getTalentRank(TALENTS.INSIDIOUS_IRE_TALENT) *
+    PHANTASMAL_PATHOGEN_DAMAGE_PER_RANK;
+
+  constructor(options: Options) {
+    super(options);
+    this.addEventListener(
+      Events.cast.by(SELECTED_PLAYER).spell(SPELLS.SHADOWY_APPARITION_CAST),
+      this.onCastSA,
+    );
+  }
+
+  onCastSA() {
+    this.castSA += 1;
+  }
+
+  phantasmalPathogen() {
+    const spell = SPELLS.SHADOWY_APPARITION_DAMAGE;
+    const damageInstances = this.eventHistory.getEvents(EventType.Damage, {
+      spell,
+    });
+
+    const [pathogen, unpathogen] = damageInstances.reduce<[DamageEvent[], DamageEvent[]]>(
+      (result, damage) => {
+        const enemy = this.enemies.getEntity(damage);
+        const wasDP = enemy && enemy.hasBuff(TALENTS.DEVOURING_PLAGUE_TALENT.id, damage.timestamp);
+        result[wasDP ? 0 : 1].push(damage);
+        return result;
+      },
+      [[], []],
+    );
```

Comment:
this is going to be quadratic complexity and is very likely to perform super bad in M+.

rather than using `getEvents` + reducing w/ `hasBuff(x, timestamp)`, use event listeners to collect the damage + buff status (using `hasBuff` without a specific timestamp is plenty fast if you just want to call it inside a shadowy apparition damage event handler).

Reply:
That is good to know.  I have changed it to use vent listeners instead of eventHistory.